File: next_smaller_num_py/next_smaller.py
def next_smaller(n):
    s = str(n)
    num = None

    if s == "".join(sorted(s)):
        return -1

    for i in range(len(s) - 2, -1, -1):
        if s[i] > s[i + 1]:
            for j in range(len(s) - 1, i, -1):
                if s[i] > s[j]:
                    num = int(s[:i] + s[j] + s[-1:j:-1] + s[i] + s[j - 1 : i : -1])
                    break
            break

    return -1 if len(str(num)) < len(s) else num

File: javascript/next_smaller_num/next_smaller.js
function nextSmaller(n) {
  const numArr = Array.from(String(n), Number)
  const sortedArr = Array.from(String(n), Number).sort()

  if (numArr.join('') === sortedArr.join('')) return -1

  let i, prev = 9
  for (i = numArr.length; i--;) {
    if (numArr[i] > prev) break
    prev = numArr[i]
  }

  const pivotI = i
  const pivot = numArr[pivotI]

  for (i = numArr.length; i--;) {
    if (numArr[i] < pivot) break
  }

  numArr[pivotI] = numArr[i]
  numArr[i] = pivot

  if (numArr[0] === 0) return -1

  return +numArr.slice(0, pivotI + 1).concat(numArr.slice(pivotI + 1).reverse()).join('')
}
module.exports = nextSmaller;
